### PR TITLE
[MNT] Unit testing revamp part 1: `check_estimator`

### PR DESCRIPTION
**Pull request by MatthewMiddlehurst**
#### Reference Issues/PRs

Relies on #1476 
See [AEP 05](https://github.com/aeon-toolkit/aeon-admin/pull/10/files)

#### What does this implement/fix? Explain your changes.

Updates the `check_estimator` method to use a scikit-learn like method for generating estimator tests rather than the current `pytest` fixtures. The old method is still available for now in `check_estimator_legacy`.  `parametrize_with_checks` simply outputs the generated checks as parameters for a `pytest` function.

`_yield_estimator_checks.py` ports tests over from `test_all_estimators`. The data used and the calling of estimator methods (i.e. `fit` and `predict`) has to change due to the detachment from the "scenarios" framework used previously.

Quick run down of contents:
- Has some small changes to base classes mostly regarding meta-data, as the storing of it in class attributes breaks the condition of no changes in `predict`/`transform`.
- Implements `parametrize_with_checks` and `check_estimator` for non-deprecated interfaces. These are only run on test estimators for now.
- Implements the current `BaseEstimator` tests as checks for the above without the "scenario" framework.

### CODE REVIEW

File: aeon/base/_base.py [hunk 1]
```diff
@@ -485,46 +485,44 @@ def get_test_params(cls, parameter_set="default"):
...
+    def create_test_instance(cls, parameter_set="default", return_first=True):
         """
         Construct Estimator instance if possible.
 
+        Calls the `get_test_params` method and returns an instance or list of instances
+        using the returned dict or list of dict.
+
         Parameters
         ----------
         parameter_set : str, default="default"
```
Comment (baraline) [inline aeon/base/_base.py:497]: Missing `cls` docstring

File: aeon/base/_base_collection.py [hunk 2]
```diff
@@ -43,7 +43,7 @@ def __init__(self):
         super().__init__()
         _check_estimator_deps(self)
 
-    def _preprocess_collection(self, X):
+    def _preprocess_collection(self, X, store_metadata=True):
         """Preprocess input X prior to call to fit.
 
         1. Checks the characteristics of X, store metadata, checks self can handle
```
Comment (baraline) [inline aeon/base/_base_collection.py:49]: Missing `store_metadata`  doc string

File: aeon/base/_base.py (hunk 1 above)
Comment (baraline) [inline aeon/base/_base.py:497]: Missing `cls` docstring

File: aeon/testing/estimator_checking/_estimator_checks.py [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+    independently and report the checks that are failing.
+
+    scikit-learn provides a pytest specific decorator,
+    :func:`~sklearn.utils.estimator_checks.parametrize_with_checks`, making it
+    easier to test multiple estimators.
+
+    Parameters
+    ----------
+    estimator : BaseEstimator instance or classe
+        Estimator instance or class to check.
```
Comment (baraline) [inline aeon/testing/estimator_checking/_estimator_checks.py:164]: Missing other parameter docstring, as this is a core function of the new testing module I feel like we would need the full docstring here

**Review by baraline — COMMENTED** [review]
Had a quick look and I must say that the testing pipeline seems way clearer to me with this format, from testing data to which test are run for which estimator. 

I'll delve a bit more into it tomorrow, just some typos in found while reading through it.

File: aeon/base/_base.py (hunk 1 above)
Comment (MatthewMiddlehurst) [reply aeon/base/_base.py:497]: `cls` is just the class, like `self`.

File: aeon/testing/estimator_checking/_estimator_checks.py (hunk 3 above)
Comment (MatthewMiddlehurst) [reply aeon/testing/estimator_checking/_estimator_checks.py:164]: The docstring for both of these methods are actually old, I forgot to update them! Good catch, will get it in next commit.

**Comment by MatthewMiddlehurst** [comment]
Actually, with forecasting more officially deprecated now, I can probably remove the legacy check function. Can keep for now if people disagree.

**Comment by TonyBagnall** [comment]
> Actually, with forecasting more officially deprecated now, I can probably remove the legacy check function. Can keep for now if people disagree.

I vote remove it

File: aeon/testing/estimator_checking/_estimator_checks.py [hunk 4]
```diff
@@ -0,0 +1,411 @@
...
+    estimator: Union[BaseEstimator, Type[BaseEstimator]],
+    raise_exceptions: bool = False,
+    use_first_parameter_set: bool = False,
+    checks_to_run: Union[str, List[str]] = None,
+    checks_to_exclude: Union[str, List[str]] = None,
+    full_checks_to_run: Union[str, List[str]] = None,
+    full_checks_to_exclude: Union[str, List[str]] = None,
+    verbose: bool = False,
+):
+    """Check if estimator adheres to scikit-learn conventions.
```
Comment (TonyBagnall) [inline aeon/testing/estimator_checking/_estimator_checks.py:143]: aeon rather than scikit?

File: aeon/transformations/series/base.py [hunk 5]
```diff
@@ -186,9 +184,7 @@ def inverse_transform(self, X, y=None, axis=1):
 
         # check whether is fitted
         self.check_is_fitted()
-        X = self._preprocess_series(
-            X, axis=axis, store_metadata=self.get_class_tag("fit_is_empty")
-        )
+        X = self._preprocess_series(X, axis=axis, store_metadata=False)
```
Comment (TonyBagnall) [inline aeon/transformations/series/base.py:187]: if store metadata is now false everywhere, I suggest remove it

Comment (TonyBagnall) [reply aeon/transformations/series/base.py:187]: but its not, its false in some places, need to look closer

File: aeon/testing/estimator_checking/_estimator_checks.py (hunk 4 above)
Comment (MatthewMiddlehurst) [reply aeon/testing/estimator_checking/_estimator_checks.py:143]: good catch 🙂

**Review by TonyBagnall — APPROVED** [review]
LGTM, onwards!

**Review by baraline — APPROVED** [review]

**Pull request merged**